src/workerd/io/trace-stream.c++

```diff
@@ -0,0 +1,937 @@
+  obj.set(js, SPANID_STR, js.str(event.spanId.toGoString()));
+  obj.set(js, TIMESTAMP_STR, js.date(event.timestamp));
+  obj.set(js, SEQUENCE_STR, js.num(event.sequence));
+
+  KJ_SWITCH_ONEOF(event.event) {
+    KJ_CASE_ONEOF(onset, tracing::Onset) {
+      obj.set(js, EVENT_STR, ToJs(js, onset, cache));
+    }
+    KJ_CASE_ONEOF(outcome, tracing::Outcome) {
+      obj.set(js, EVENT_STR, ToJs(js, outcome, cache));
+    }
+    KJ_CASE_ONEOF(hibernate, tracing::Hibernate) {
+      obj.set(js, EVENT_STR, ToJs(js, hibernate, cache));
+    }
+    KJ_CASE_ONEOF(spanOpen, tracing::SpanOpen) {
+      obj.set(js, EVENT_STR, ToJs(js, spanOpen, cache));
+    }
+    KJ_CASE_ONEOF(spanClose, tracing::SpanClose) {
+      obj.set(js, EVENT_STR, ToJs(js, spanClose, cache));
+    }
+    KJ_CASE_ONEOF(mark, tracing::Mark) {
+      KJ_SWITCH_ONEOF(mark) {
+        KJ_CASE_ONEOF(de, tracing::DiagnosticChannelEvent) {
+          obj.set(js, EVENT_STR, ToJs(js, de, cache));
+        }
+        KJ_CASE_ONEOF(ex, tracing::Exception) {
+          obj.set(js, EVENT_STR, ToJs(js, ex, cache));
+        }
+        KJ_CASE_ONEOF(log, tracing::Log) {
+          obj.set(js, EVENT_STR, ToJs(js, log, cache));
+        }
+        KJ_CASE_ONEOF(ret, tracing::Return) {
+          obj.set(js, EVENT_STR, ToJs(js, ret, cache));
+        }
+        KJ_CASE_ONEOF(link, tracing::Link) {
+          obj.set(js, EVENT_STR, ToJs(js, link, cache));
+        }
+        KJ_CASE_ONEOF(attrs, kj::Array<tracing::Attribute>) {
+          obj.set(js, EVENT_STR, ToJs(js, attrs, cache));
+        }
+      }
+    }
+  }
+
+  return obj;
+}
+
+class ServerTopLevelMembrane final: public capnp::MembranePolicy, public kj::Refcounted {
+ public:
+  explicit ServerTopLevelMembrane(kj::Own<kj::PromiseFulfiller<void>> doneFulfiller)
+      : doneFulfiller(kj::mv(doneFulfiller)) {}
+  ~ServerTopLevelMembrane() noexcept(false) {
+    KJ_IF_SOME(f, doneFulfiller) {
+      f->reject(
+          KJ_EXCEPTION(DISCONNECTED, "Tail stream session canceled without handling any events."));
+    }
+  }
+
+  kj::Maybe<capnp::Capability::Client> inboundCall(
+      uint64_t interfaceId, uint16_t methodId, capnp::Capability::Client target) override {
+    auto f = kj::mv(JSG_REQUIRE_NONNULL(
+        doneFulfiller, Error, "Only one tailStream method call is allowed on this object."));
+    doneFulfiller = kj::none;
+    return capnp::membrane(kj::mv(target), kj::refcounted<CompletionMembrane>(kj::mv(f)));
+  }
+
+  kj::Maybe<capnp::Capability::Client> outboundCall(
+      uint64_t interfaceId, uint16_t methodId, capnp::Capability::Client target) override {
+    KJ_FAIL_ASSERT("ServerTopLevelMembrane shouldn't have outgoing capabilities");
+  }
+
+  kj::Own<MembranePolicy> addRef() override {
+    return kj::addRef(*this);
+  }
+
+ private:
+  kj::Maybe<kj::Own<kj::PromiseFulfiller<void>>> doneFulfiller;
+};
+
+class TailStreamTargetBase: public rpc::TailStreamTarget::Server {
```

Comment:
Review note: The TailStreamTargetBase forms the basis of the server-side of the tail stream RPC interface. It arranges to have the received tail events dispatched to the correct exported handler